File: smarttrade-web/src/features/trading-journal/data/repositories/journalRepositoryImpl.ts
import { DatabaseFailure } from '@/core/error/failure'
import { err, ok, type Result } from '@/core/result/result'
import type { JournalEntry } from '../../domain/entities/journalEntry'
import type { JournalRepository } from '../../domain/repositories/journalRepository'
import { journalLocalDataSource } from '../datasources/journalLocalDataSource'

export class JournalRepositoryImpl implements JournalRepository {
  async getEntries(): Promise<Result<JournalEntry[]>> {
    try {
      return ok(await journalLocalDataSource.getEntries())
    } catch (error) {
      return err(new DatabaseFailure(`Falha ao carregar o diário: ${error}`))
    }
  }

  async addEntry(entry: JournalEntry): Promise<Result<void>> {
    try {
      await journalLocalDataSource.addEntry(entry)
      return ok(undefined)
    } catch (error) {
      return err(new DatabaseFailure(`Falha ao salvar a operação: ${error}`))
    }
  }

  async deleteEntry(id: string): Promise<Result<void>> {
    try {
      await journalLocalDataSource.deleteEntry(id)
      return ok(undefined)
    } catch (error) {
      return err(new DatabaseFailure(`Falha ao remover a operação: ${error}`))
    }
  }
}

export const journalRepository: JournalRepository = new JournalRepositoryImpl()

File: smarttrade_app/lib/features/trading_journal/data/repositories/journal_repository_impl.dart
import 'package:dartz/dartz.dart';
import 'package:smarttrade_app/core/error/failures.dart';
import 'package:smarttrade_app/features/trading_journal/data/datasources/journal_local_datasource.dart';
import 'package:smarttrade_app/features/trading_journal/domain/entities/journal_entry.dart';
import 'package:smarttrade_app/features/trading_journal/domain/repositories/journal_repository.dart';

class JournalRepositoryImpl implements JournalRepository {
  const JournalRepositoryImpl(this._localDataSource);

  final JournalLocalDataSource _localDataSource;

  @override
  Future<Either<Failure, List<JournalEntry>>> getEntries() async {
    try {
      return Right(await _localDataSource.getEntries());
    } catch (error) {
      return Left(DatabaseFailure('Falha ao carregar o diário: $error'));
    }
  }

  @override
  Future<Either<Failure, Unit>> addEntry(JournalEntry entry) async {
    try {
      await _localDataSource.addEntry(entry);
      return const Right(unit);
    } catch (error) {
      return Left(DatabaseFailure('Falha ao salvar a operação: $error'));
    }
  }

  @override
  Future<Either<Failure, Unit>> deleteEntry(String id) async {
    try {
      await _localDataSource.deleteEntry(id);
      return const Right(unit);
    } catch (error) {
      return Left(DatabaseFailure('Falha ao remover a operação: $error'));
    }
  }
}
